receiver/prometheusreceiver/internal/staleness_store.go

```diff
@@ -66,24 +78,40 @@ func (ss *stalenessStore) refresh() {
 
 // isStale returns whether lbl was seen only in the previous scrape and not the current.
 func (ss *stalenessStore) isStale(lbl labels.Labels) bool {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+
 	hash := lbl.Hash()
-	return ss.previousHashes[hash] && !ss.currentHashes[hash]
+	_, inPrev := ss.previousHashes[hash]
+	_, inCurrent := ss.currentHashes[hash]
+	return inPrev && !inCurrent
 }
 
 // markAsCurrentlySeen adds lbl to the manifest of labels seen in the current scrape.
 // This method should be called before refresh, but during a scrape whenever labels are encountered.
-func (ss *stalenessStore) markAsCurrentlySeen(lbl labels.Labels) {
-	ss.currentHashes[lbl.Hash()] = true
+func (ss *stalenessStore) markAsCurrentlySeen(lbl labels.Labels, at int64) {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+
+	ss.currentHashes[lbl.Hash()] = at
 	ss.current = append(ss.current, lbl)
 }
 
+type staleEntry struct {
+	labels labels.Labels
+	at     int64
+}
+
 // emitStaleLabels returns the labels that were previously seen in
 // the prior scrape, but are not currently present in this scrape cycle.
-func (ss *stalenessStore) emitStaleLabels() (stale []labels.Labels) {
+func (ss *stalenessStore) emitStaleLabels() (stale []*staleEntry) {
```

Comment:
Nit: this possibly will have better performance due to lower number of allocations.
```suggestion
func (ss *stalenessStore) emitStaleLabels() (stale []staleEntry) {
```

Reply:
The opposite happens though, as these entries are moved around, a pointer is kept around while values will be copied, which will cause more allocations.

Reply:
I was curious so did a quick benchmark :-)

```go
type staleEntry struct {
	labels   labels.Labels
	seenAtMs int64
}

func emitStaleLabelsByPtr() (stale []*staleEntry) {
	var labels labels.Labels
	for i := 0; i < 10; i++ {
		stale = append(stale, &staleEntry{seenAtMs: int64(i), labels: labels})
	}
	return stale
}

func emitStaleLabelsByVal() (stale []staleEntry) {
	var labels labels.Labels
	for i := 0; i < 10; i++ {
		stale = append(stale, staleEntry{seenAtMs: int64(i), labels: labels})
	}
	return stale
}

func BenchmarkByPtr(b *testing.B) {
	for i := 0; i < b.N; i++ {
		emitStaleLabelsByPtr()
	}
}

func BenchmarkByVal(b *testing.B) {
	for i := 0; i < b.N; i++ {
		emitStaleLabelsByVal()
	}
}
```

Results:
```
$ go test -bench . --benchmem
goos: darwin
goarch: amd64
pkg: github.com/tigrannajaryan/gotest
cpu: Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
BenchmarkByPtr-8         2093342               618.8 ns/op           568 B/op         15 allocs/op
BenchmarkByVal-8         2512652               403.6 ns/op           992 B/op          5 allocs/op
PASS
ok      github.com/tigrannajaryan/gotest        5.652s
```

I tried to approximate the code in this PR as best as I could while keeping it minimal, so hopefully it the results are representative.

It appears by value is a bit faster with less allocations, but the difference is negligible unless it is in some hot path.

Reply:
Thanks @tigrannajaryan! The copying that happens is when for example any of those values is passed into a function or moved into a new scope; by value they'll ALWAYS be copied byte by byte which is something we don't want to happen and cause memory bloat. Actually one time while doing an inspection I noticed that the collector has a whole bunch of these patterns e.g.
```go
package main

type st struct {
   id int
   name string
   bloat [1024]byte
}

func foo(s st) {
    // Do something with s.
    ...
}

func fooPtr(s *st) {
    // Do something with s.
    ...
}

func main() {
     sl := []st{{id: 0, name: "a"}, {id: 1, name: "b"}}
    for i := 0; i < 10; i++ {
          for _, s := range sl {
             foo(s) // Copy byte by byte will ALWAYS happen here.. If the struct is big, lots of memory bloat.
          }
     }

    slp := []*st{{id: 0, name: "a"}, {id: 1, name: "b"}}
    for i := 0; i < 10; i++ {
          for _, s := range slp {
              fooPtr(s) // No need to copy over memory here
          }
     }
}
```

Reply:
Yes, it depends on size of the struct. For sufficiently small structs copying can overall be faster since individual allocations are expensive. We see it in the benchmark I posted.